File: ch08/sec8.8/foundations/forward_pass.c
/*
 * forward_pass.c
 *
 * Compile:  gcc -O2 -o forward_pass forward_pass.c -lm
 * Run:      ./forward_pass
 *
 * Implements the forward pass for a small feedforward network
 * with architecture [2, 3, 1] entirely from first principles.
 *
 * This mirrors the Python version but makes the raw arithmetic
 * explicit: there is no framework, no autograd, no abstraction.
 * Each weight multiplication and addition is written out.
 *
 * Layer dimensions:
 *   Input:   2 units
 *   Hidden:  3 units  (ReLU activation)
 *   Output:  1 unit   (Sigmoid activation)
 */

#include <math.h>
#include <stdio.h>

#define INPUT_DIM  2
#define HIDDEN_DIM 3
#define OUTPUT_DIM 1

static double relu(double z) {
    return z > 0.0 ? z : 0.0;
}

static double sigmoid(double z) {
    return 1.0 / (1.0 + exp(-z));
}

static void matmul_add(
    const double *W, const double *b,
    const double *h_in, double *z_out,
    int rows, int cols)
{
    for (int i = 0; i < rows; i++) {
        double sum = b[i];
        for (int j = 0; j < cols; j++) {
            sum += W[i * cols + j] * h_in[j];
        }
        z_out[i] = sum;
    }
}

static void apply_activation(
    const double *z, double *h, int n,
    double (*act)(double))
{
    for (int i = 0; i < n; i++) {
        h[i] = act(z[i]);
    }
}

static double bce_loss(double y_hat, double y) {
    double eps = 1e-9;
    return -(y * log(y_hat + eps) + (1.0 - y) * log(1.0 - y_hat + eps));
}

int main(void) {
    /*
     * Weights for hidden layer: shape [HIDDEN_DIM, INPUT_DIM]
     * Stored in row-major order. W1[i][j] = W1[i * INPUT_DIM + j]
     */
    double W1[HIDDEN_DIM * INPUT_DIM] = {
         0.5, -0.2,
        -0.3,  0.8,
         0.1,  0.4
    };
    double b1[HIDDEN_DIM] = { 0.0, 0.1, -0.1 };

    /* Weights for output layer: shape [OUTPUT_DIM, HIDDEN_DIM] */
    double W2[OUTPUT_DIM * HIDDEN_DIM] = { 0.6, -0.5, 0.3 };
    double b2[OUTPUT_DIM] = { 0.05 };

    double x[INPUT_DIM] = { 0.5, -0.3 };
    double y = 1.0;

    /* Hidden layer */
    double z1[HIDDEN_DIM];
    double h1[HIDDEN_DIM];

    matmul_add(W1, b1, x, z1, HIDDEN_DIM, INPUT_DIM);
    apply_activation(z1, h1, HIDDEN_DIM, relu);

    /* Output layer */
    double z2[OUTPUT_DIM];
    double h2[OUTPUT_DIM];

    matmul_add(W2, b2, h1, z2, OUTPUT_DIM, HIDDEN_DIM);
    apply_activation(z2, h2, OUTPUT_DIM, sigmoid);

    double loss = bce_loss(h2[0], y);

    printf("Input:          [%.3f, %.3f]\n", x[0], x[1]);
    printf("z1 (pre-act):   [%.4f, %.4f, %.4f]\n", z1[0], z1[1], z1[2]);
    printf("h1 (after ReLU):[%.4f, %.4f, %.4f]\n", h1[0], h1[1], h1[2]);
    printf("z2 (pre-act):   [%.4f]\n", z2[0]);
    printf("y_hat (sigmoid):[%.6f]\n", h2[0]);
    printf("Target y:        %.1f\n", y);
    printf("BCE loss:        %.6f\n", loss);

    return 0;
}

File: ch08/sec8.8/foundations/forward_pass.py
"""
forward_pass.py

Implements a minimal feedforward neural network using only NumPy.
The network has a fixed architecture: 2 -> 4 -> 4 -> 1.

The chapter describes the layer computation as:
    z^(l) = W^(l) h^(l-1) + b^(l)
    h^(l) = sigma(z^(l))

This file makes each of those steps explicit and prints intermediate
values so you can follow the data through the network by hand.
"""

import numpy as np


def relu(z):
    return np.maximum(0.0, z)


def relu_derivative(z):
    return (z > 0).astype(float)


def sigmoid(z):
    return 1.0 / (1.0 + np.exp(-z))


def softmax(z):
    shifted = z - np.max(z)
    exp_z = np.exp(shifted)
    return exp_z / np.sum(exp_z)


def initialise_parameters(layer_dims, rng):
    """
    Xavier (Glorot) initialisation: scale weights by 1/sqrt(fan_in).
    Biases are initialised to zero.

    Parameters
    ----------
    layer_dims : list of int
        Number of units in each layer, including the input layer.
    rng : np.random.Generator

    Returns
    -------
    params : list of (W, b) tuples, one per layer transition.
    """
    params = []
    for i in range(1, len(layer_dims)):
        fan_in = layer_dims[i - 1]
        fan_out = layer_dims[i]
        scale = np.sqrt(1.0 / fan_in)
        W = rng.normal(0.0, scale, size=(fan_out, fan_in))
        b = np.zeros((fan_out, 1))
        params.append((W, b))
    return params


def forward_pass(x, params, activation=relu, output_activation=sigmoid):
    """
    Runs a full forward pass through the network.

    Parameters
    ----------
    x : np.ndarray, shape (d, 1)
        Input column vector.
    params : list of (W, b) tuples
    activation : callable
        Hidden layer activation function.
    output_activation : callable
        Output layer activation function.

    Returns
    -------
    activations : list of np.ndarray
        The activation h^(l) at every layer, including the input (l=0).
    pre_activations : list of np.ndarray
        The pre-activation z^(l) at every layer (excluding the input).
    """
    h = x
    activations = [h]
    pre_activations = []

    for layer_idx, (W, b) in enumerate(params):
        z = W @ h + b
        pre_activations.append(z)

        is_last_layer = (layer_idx == len(params) - 1)
        if is_last_layer:
            h = output_activation(z)
        else:
            h = activation(z)

        activations.append(h)

    return activations, pre_activations


def binary_cross_entropy(y_hat, y):
    eps = 1e-9
    return -(y * np.log(y_hat + eps) + (1 - y) * np.log(1 - y_hat + eps))


def print_pass(x, y, params):
    print("Input x:")
    print(x.T)

    activations, pre_activations = forward_pass(x, params)

    for l, (z, h) in enumerate(zip(pre_activations, activations[1:]), start=1):
        print(f"\nLayer {l}")
        print(f"  pre-activation z^({l}):  {z.T}")
        print(f"  activation    h^({l}):  {h.T}")

    y_hat = activations[-1]
    loss = binary_cross_entropy(y_hat, y)
    print(f"\nTarget y       : {y}")
    print(f"Prediction y_hat: {float(y_hat.flat[0]):.6f}")
    print(f"Binary cross-entropy loss: {float(loss.flat[0]):.6f}")


if __name__ == "__main__":
    rng = np.random.default_rng(seed=42)

    layer_dims = [2, 4, 4, 1]
    params = initialise_parameters(layer_dims, rng)

    x = np.array([[0.5], [-0.3]])
    y = 1.0

    print_pass(x, y, params)
